util/intensity_explorer.py

```diff
@@ -0,0 +1,623 @@
+          var_weight = 1 / (sum_weight - sum_square_weights[-1] / sum_weight)
+          variances.extend(
+              flex.double(count, weighted_sum_square_residual * var_weight))
+        # Record the variances in the reflection table.
+        rtable['intensity.mean.variance'] = variances
+
+      self.rtables[key] = rtable
+
+  def _make_z(self, uncertainty='sigma'):
+    u"""
+    Generate reflection z-scores.
+
+    Calculate z-scores from reflection intensities, weighted mean
+    intensities and a chosen measure of uncertainty in the intensity
+    measurement.
+
+    :param uncertainty: Chosen measure of uncertainty.  Options are
+      * ``stddev`` — standard deviation, as calculated from the unbiased
+      weighted variance aggregated amongst all symmetry-equivalent reflections;
+      * ``sigma`` — measurement error for individual reflections.
+    :type uncertainty: str
+    """
+
+    uncertainty_name = {
+      'sigma': 'intensity.sum.variance',
+      'stddev': 'intensity.mean.variance'
+    }[uncertainty]
+
+    for key, rtable in self.rtables.items():
+      try:
+        uncertainty_value = flex.sqrt(rtable[uncertainty_name])
+      except KeyError:
+        from logging import warn
+        uncertainty_value = flex.sqrt(rtable['intensity.sum.variance'])
+        warn(u"""Weighted variances haven't been calculated,
+      be sure to specify calculate_variances=True to use them.
+      Defaulting to measured σ values as a measure of uncertainty instead.""")
+
+      z = (rtable['intensity.sum.value']
+           - rtable['intensity.mean.value']) / uncertainty_value
+      rtable['intensity.z_score'] = z
+
+      self.rtables[key] = rtable
+
+  def _probplot_data(self):
+    """Generate the data for a normal probability plot of z-scores."""
+
+    from scipy import stats as ss
+
+    for key, rtable in self.rtables.items():
+      order = flex.sort_permutation(rtable['intensity.z_score'])
+      osm = flex.double(rtable.size(), 0)
+      osm.set_selected(order,
+                       flex.double(ss.probplot(rtable['intensity.z_score'],
+                                               fit=False)[0]))
+      rtable['intensity.order_statistic_medians'] = osm
+
+      self.rtables[key] = rtable
+
+  def plot_z_histogram(self):
+    """Plot a histogram of the z-scores of the weighted mean intensities."""
+
+    from matplotlib import pyplot as plt
+
+    fig, ax = plt.subplots()
+
+    ax.set_title(r'$z$ histogram')
+    ax.set_xlabel(r'$z$')
+    ax.set_ylabel(r'$N$')
+    ax.hist(self.rtable['intensity.z_score'],
+            label='$z$', bins=100, range=(-5, 5))
+    fig.savefig(self.outfile + '_zhistogram', transparent=True)
+    plt.close()
+
+  def _plot_symmetry_equivalents(self,
+                                 overlay_mean=False,
+                                 overlay_uncertainty=False):
+    """Really just a test plot.  Slow.  You probably don't want to use."""
+
+    from matplotlib import pyplot as plt
```

Comment:
import used multiple times. Consider moving to top, leaving out the ```as``` bit and using ```pyplot``` throughout.

Reply:
I thought you might say that.  I was trying to keep an expensive import out of the way because the methods that use `pyplot` are unlikely to get much use.  In light of that, a better arrangement is to remove these methods altogether.

Reply:
I see. Maybe bundle the mathplotlib-depedent methods into a visualization class in a separate file that takes an ```IntensityDist```?

Reply:
Exactly.  Currently living in dials_scratch while I do something else.